From 0f5041ce1fa1e5ec401050a7bf3374c3317dfbd3 Mon Sep 17 00:00:00 2001
From: Alison Wang <alison.wang@nxp.com>
Date: Thu, 11 Jul 2019 15:10:53 +0800
Subject: [PATCH] Enable POLICY_OTA for  Layerscape boards

This patch enables POLICY_OTA for ls1012afrwy, ls1012ardb, ls1028ardb,
ls1043ardb, ls1046afrwy, ls1046ardb, ls1088ardb, ls2088ardb and
lx2160ardb.

Signed-off-by: Alison Wang <alison.wang@nxp.com>
---
 arch/arm/cpu/armv8/fsl-layerscape/Kconfig | 16 +++++++++++++
 arch/arm/cpu/armv8/fsl-layerscape/cpu.c   |  4 ++++
 board/freescale/ls1012afrdm/ls1012afrdm.c |  1 +
 board/freescale/ls1012ardb/ls1012ardb.c   |  1 +
 board/freescale/ls1043ardb/ls1043ardb.c   |  1 +
 board/freescale/ls1046afrwy/ls1046afrwy.c |  1 +
 board/freescale/ls1046ardb/ls1046ardb.c   |  1 +
 board/freescale/ls1088a/ls1088a.c         |  1 +
 board/freescale/ls2080ardb/ls2080ardb.c   |  1 +
 board/freescale/lx2160a/lx2160a.c         |  1 +
 configs/ls1012afrwy_tfa_defconfig         |  2 ++
 configs/ls1012ardb_tfa_defconfig          |  2 ++
 configs/ls1028ardb_tfa_defconfig          |  1 +
 configs/ls1043ardb_tfa_defconfig          |  1 +
 configs/ls1046afrwy_tfa_defconfig         |  1 +
 configs/ls1046ardb_tfa_defconfig          |  1 +
 configs/ls1088ardb_tfa_defconfig          |  1 +
 configs/ls2088ardb_tfa_defconfig          |  1 +
 configs/lx2160ardb_tfa_defconfig          |  1 +
 drivers/net/pfe_eth/pfe_firmware.c        | 38 +++++++++++++++++++++++++------
 include/configs/ls1012afrwy.h             |  9 ++++++++
 include/configs/ls1012ardb.h              |  8 +++++++
 include/configs/ls1046afrwy.h             |  3 +++
 23 files changed, 90 insertions(+), 7 deletions(-)

diff --git a/arch/arm/cpu/armv8/fsl-layerscape/Kconfig b/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
index 8c47d81..918bfcb 100644
--- a/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
+++ b/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
@@ -627,3 +627,19 @@ config TFABOOT
        help
          Enabling this will make a U-Boot binary that is capable of being
          booted via TFA.
+
+config POLICY_OTA
+       bool "Support OTA"
+       default n
+       help
+         Enabling this will make a U-Boot binary that supports OTA.
+
+config SYS_PFE_FW_LENGTH
+       bool "PFE firmware length"
+       depends on ARCH_LS1012A && POLICY_OTA
+       default n
+
+config SYS_PFE_FW_ADDR
+       bool "PFE firmware address"
+       depends on ARCH_LS1012A && POLICY_OTA
+       default n
diff --git a/arch/arm/cpu/armv8/fsl-layerscape/cpu.c b/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
index b8fa736..e6473e4 100644
--- a/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
+++ b/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
@@ -737,6 +737,10 @@ enum boot_src __get_boot_src(u32 porsr1)
 	if (CONFIG_IS_ENABLED(SYS_FSL_ERRATUM_A010539) && !rcw_src)
 		src = BOOT_SOURCE_QSPI_NOR;
 
+#ifdef	CONFIG_POLICY_OTA
+	src = BOOT_SOURCE_SD_MMC;
+#endif
+
 	debug("%s: src 0x%x\n", __func__, src);
 	return src;
 }
diff --git a/board/freescale/ls1012afrdm/ls1012afrdm.c b/board/freescale/ls1012afrdm/ls1012afrdm.c
index b4c611e..5b63859 100644
--- a/board/freescale/ls1012afrdm/ls1012afrdm.c
+++ b/board/freescale/ls1012afrdm/ls1012afrdm.c
@@ -61,6 +61,7 @@ int checkboard(void)
 		break;
 	}
 #endif
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/ls1012ardb/ls1012ardb.c b/board/freescale/ls1012ardb/ls1012ardb.c
index f648a90..f4b0155 100644
--- a/board/freescale/ls1012ardb/ls1012ardb.c
+++ b/board/freescale/ls1012ardb/ls1012ardb.c
@@ -84,6 +84,7 @@ int checkboard(void)
 
 	puts("Board: LS1012A2G5RDB ");
 #endif
+	printf("EdgeScale Bootstrap U-Boot\n");
 	return 0;
 }
 
diff --git a/board/freescale/ls1043ardb/ls1043ardb.c b/board/freescale/ls1043ardb/ls1043ardb.c
index fbd9a26..d5ab5c2 100644
--- a/board/freescale/ls1043ardb/ls1043ardb.c
+++ b/board/freescale/ls1043ardb/ls1043ardb.c
@@ -180,6 +180,7 @@ int checkboard(void)
 	puts("SERDES Reference Clocks:\n");
 	sd1refclk_sel = CPLD_READ(sd1refclk_sel);
 	printf("SD1_CLK1 = %s, SD1_CLK2 = %s\n", freq[sd1refclk_sel], freq[0]);
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/ls1046afrwy/ls1046afrwy.c b/board/freescale/ls1046afrwy/ls1046afrwy.c
index ed64ac3..2b84e91 100644
--- a/board/freescale/ls1046afrwy/ls1046afrwy.c
+++ b/board/freescale/ls1046afrwy/ls1046afrwy.c
@@ -128,6 +128,7 @@ int checkboard(void)
 	else
 		puts("QSPI\n");
 	printf("SD1_CLK1 = %s, SD1_CLK2 = %s\n", freq[0], freq[1]);
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/ls1046ardb/ls1046ardb.c b/board/freescale/ls1046ardb/ls1046ardb.c
index 0a73fe8..3fb463a 100644
--- a/board/freescale/ls1046ardb/ls1046ardb.c
+++ b/board/freescale/ls1046ardb/ls1046ardb.c
@@ -61,6 +61,7 @@ int checkboard(void)
 	puts("SERDES Reference Clocks:\n");
 	sd1refclk_sel = CPLD_READ(sd1refclk_sel);
 	printf("SD1_CLK1 = %s, SD1_CLK2 = %s\n", freq[sd1refclk_sel], freq[0]);
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/ls1088a/ls1088a.c b/board/freescale/ls1088a/ls1088a.c
index 96f0871..ab3bbb5 100644
--- a/board/freescale/ls1088a/ls1088a.c
+++ b/board/freescale/ls1088a/ls1088a.c
@@ -311,6 +311,7 @@ int checkboard(void)
 	printf("Clock1 = %sMHz ", freq[clock]);
 	clock = (sw >> 0) & 3;
 	printf("Clock2 = %sMHz\n", freq[clock]);
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/ls2080ardb/ls2080ardb.c b/board/freescale/ls2080ardb/ls2080ardb.c
index 3af83f9..a21d959 100644
--- a/board/freescale/ls2080ardb/ls2080ardb.c
+++ b/board/freescale/ls2080ardb/ls2080ardb.c
@@ -130,6 +130,7 @@ int checkboard(void)
 	puts("\nSERDES2 Reference : ");
 	printf("Clock1 = 100MHz ");
 	printf("Clock2 = 100MHz\n");
+	printf("EdgeScale Bootstrap U-Boot\n");
 
 	return 0;
 }
diff --git a/board/freescale/lx2160a/lx2160a.c b/board/freescale/lx2160a/lx2160a.c
index 9f9fb79..16a528d 100644
--- a/board/freescale/lx2160a/lx2160a.c
+++ b/board/freescale/lx2160a/lx2160a.c
@@ -359,6 +359,7 @@ int checkboard(void)
 	puts("SERDES2 Reference: Clock1 = 100MHz Clock2 = 100MHz\n");
 	puts("SERDES3 Reference: Clock1 = 100MHz Clock2 = 100Hz\n");
 #endif
+	printf("EdgeScale Bootstrap U-Boot\n");
 	return 0;
 }
 
diff --git a/configs/ls1012afrwy_tfa_defconfig b/configs/ls1012afrwy_tfa_defconfig
index 3b3b0cc..3ebf217 100644
--- a/configs/ls1012afrwy_tfa_defconfig
+++ b/configs/ls1012afrwy_tfa_defconfig
@@ -27,6 +27,7 @@ CONFIG_CMD_USB=y
 CONFIG_CMD_CACHE=y
 CONFIG_OF_CONTROL=y
 CONFIG_DEFAULT_DEVICE_TREE="fsl-ls1012a-frwy"
+CONFIG_ENV_IS_IN_MMC=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_NET_RANDOM_ETHADDR=y
 CONFIG_DM=y
@@ -52,3 +53,4 @@ CONFIG_USB=y
 CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1012ardb_tfa_defconfig b/configs/ls1012ardb_tfa_defconfig
index f385d4e..b52ca195 100644
--- a/configs/ls1012ardb_tfa_defconfig
+++ b/configs/ls1012ardb_tfa_defconfig
@@ -29,6 +29,7 @@ CONFIG_CMD_USB=y
 CONFIG_CMD_CACHE=y
 CONFIG_OF_CONTROL=y
 CONFIG_DEFAULT_DEVICE_TREE="fsl-ls1012a-rdb"
+CONFIG_ENV_IS_IN_MMC=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_NET_RANDOM_ETHADDR=y
 CONFIG_DM=y
@@ -56,3 +57,4 @@ CONFIG_USB=y
 CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1028ardb_tfa_defconfig b/configs/ls1028ardb_tfa_defconfig
index 2e98e94..6f69466 100644
--- a/configs/ls1028ardb_tfa_defconfig
+++ b/configs/ls1028ardb_tfa_defconfig
@@ -63,3 +63,4 @@ CONFIG_USB_XHCI_DWC3=y
 CONFIG_WDT=y
 CONFIG_WDT_SP805=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1043ardb_tfa_defconfig b/configs/ls1043ardb_tfa_defconfig
index 5802902..0e8f201 100644
--- a/configs/ls1043ardb_tfa_defconfig
+++ b/configs/ls1043ardb_tfa_defconfig
@@ -52,3 +52,4 @@ CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1046afrwy_tfa_defconfig b/configs/ls1046afrwy_tfa_defconfig
index 36cdeb0..9b3a6d9 100644
--- a/configs/ls1046afrwy_tfa_defconfig
+++ b/configs/ls1046afrwy_tfa_defconfig
@@ -54,3 +54,4 @@ CONFIG_USB=y
 CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1046ardb_tfa_defconfig b/configs/ls1046ardb_tfa_defconfig
index 057ce9d..8f66d2a 100644
--- a/configs/ls1046ardb_tfa_defconfig
+++ b/configs/ls1046ardb_tfa_defconfig
@@ -53,3 +53,4 @@ CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls1088ardb_tfa_defconfig b/configs/ls1088ardb_tfa_defconfig
index 93fbbf3..393c7f9 100644
--- a/configs/ls1088ardb_tfa_defconfig
+++ b/configs/ls1088ardb_tfa_defconfig
@@ -62,3 +62,4 @@ CONFIG_USB_XHCI_DWC3=y
 CONFIG_USB_DWC3=y
 CONFIG_USB_GADGET=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/ls2088ardb_tfa_defconfig b/configs/ls2088ardb_tfa_defconfig
index 74fdbbe..db9ddf1 100644
--- a/configs/ls2088ardb_tfa_defconfig
+++ b/configs/ls2088ardb_tfa_defconfig
@@ -70,3 +70,4 @@ CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/configs/lx2160ardb_tfa_defconfig b/configs/lx2160ardb_tfa_defconfig
index de670c7..a901e2e 100644
--- a/configs/lx2160ardb_tfa_defconfig
+++ b/configs/lx2160ardb_tfa_defconfig
@@ -64,3 +64,4 @@ CONFIG_DM_USB=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_XHCI_DWC3=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
+CONFIG_POLICY_OTA=y
diff --git a/drivers/net/pfe_eth/pfe_firmware.c b/drivers/net/pfe_eth/pfe_firmware.c
index adb2d06..cdeba18 100644
--- a/drivers/net/pfe_eth/pfe_firmware.c
+++ b/drivers/net/pfe_eth/pfe_firmware.c
@@ -15,11 +15,13 @@
 #ifdef CONFIG_CHAIN_OF_TRUST
 #include <fsl_validate.h>
 #endif
+#ifdef CONFIG_POLICY_OTA
+#include <malloc.h>
+#include <mmc.h>
+#endif
 
 #define PFE_FIRMEWARE_FIT_CNF_NAME	"config@1"
 
-static const void *pfe_fit_addr = (void *)CONFIG_SYS_LS_PFE_FW_ADDR;
-
 /*
  * PFE elf firmware loader.
  * Loads an elf firmware image into a list of PE's (specified using a bitmask)
@@ -92,7 +94,7 @@ err:
  * @return 0 on success, a negative value on error
  */
 static int pfe_get_fw(const void **data,
-		      size_t *size, char *fw_name)
+		      size_t *size, char *fw_name, void *pfe_fit_addr)
 {
 	int conf_node_off, fw_node_off;
 	char *conf_node_name = NULL;
@@ -140,7 +142,7 @@ static int pfe_get_fw(const void **data,
  *
  * @return 0 on success, a negative value on error
  */
-static int pfe_fit_check(void)
+static int pfe_fit_check(void *pfe_fit_addr)
 {
 	int ret = 0;
 
@@ -182,14 +184,36 @@ int pfe_firmware_init(void)
 #endif
 	int ret = 0;
 	int fw_count;
+	void *pfe_fw_addr = NULL;
+
+#ifdef CONFIG_POLICY_OTA
+	int dev = CONFIG_SYS_MMC_ENV_DEV;
+
+	printf("\nPFE firmware on SD_MMC\n");
+	pfe_fw_addr = malloc(CONFIG_SYS_PFE_FW_LENGTH);
+	u32 cnt = CONFIG_SYS_PFE_FW_LENGTH / 512;
+	u32 blk = CONFIG_SYS_PFE_FW_ADDR / 512;
+	struct mmc *mmc = find_mmc_device(CONFIG_SYS_MMC_ENV_DEV);
+
+	if (!mmc) {
+		printf("\nMMC cannot find device for ucode\n");
+	} else {
+		printf("\nMMC read: dev # %u, block # %u, count %u ...\n",
+		       dev, blk, cnt);
+		mmc_init(mmc);
+		(void)blk_dread(mmc_get_blk_desc(mmc), blk, cnt, pfe_fw_addr);
+	}
+#else
+	pfe_fw_addr = (void *)CONFIG_SYS_LS_PFE_FW_ADDR;
+#endif
 
-	ret = pfe_fit_check();
+	ret = pfe_fit_check(pfe_fw_addr);
 	if (ret)
 		goto err;
 
 #ifdef CONFIG_CHAIN_OF_TRUST
 	pfe_esbc_hdr = CONFIG_SYS_LS_PFE_ESBC_ADDR;
-	pfe_img_addr = (uintptr_t)pfe_fit_addr;
+	pfe_img_addr = (uintptr_t)pfe_fw_addr;
 	if (fsl_check_boot_mode_secure() != 0) {
 		/*
 		 * In case of failure in validation, fsl_secboot_validate
@@ -214,7 +238,7 @@ int pfe_firmware_init(void)
 		else if (fw_count == 1)
 			pfe_firmware_name = "tmu";
 
-		pfe_get_fw(&raw_image_addr, &raw_image_size, pfe_firmware_name);
+		pfe_get_fw(&raw_image_addr, &raw_image_size, pfe_firmware_name, pfe_fw_addr);
 		pfe_firmware = malloc(raw_image_size);
 		if (!pfe_firmware)
 			return -ENOMEM;
diff --git a/include/configs/ls1012afrwy.h b/include/configs/ls1012afrwy.h
index 12e6437..8cd8e75 100644
--- a/include/configs/ls1012afrwy.h
+++ b/include/configs/ls1012afrwy.h
@@ -23,6 +23,15 @@
 #define CONFIG_SYS_MEMTEST_START	0x80000000
 #define CONFIG_SYS_MEMTEST_END		0x9fffffff
 
+#ifdef CONFIG_TFABOOT
+#define CONFIG_SYS_MMC_ENV_DEV         0
+
+#ifdef CONFIG_POLICY_OTA
+#define CONFIG_SYS_PFE_FW_LENGTH	0x10000
+#define CONFIG_SYS_PFE_FW_ADDR		0x20000
+#endif
+#endif
+
 #ifndef CONFIG_SPL_BUILD
 #undef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
diff --git a/include/configs/ls1012ardb.h b/include/configs/ls1012ardb.h
index f6640fa..570e1c0 100644
--- a/include/configs/ls1012ardb.h
+++ b/include/configs/ls1012ardb.h
@@ -16,6 +16,14 @@
 #define CONFIG_SYS_MEMTEST_START	0x80000000
 #define CONFIG_SYS_MEMTEST_END		0x9fffffff
 
+#ifdef CONFIG_TFABOOT
+#define CONFIG_SYS_MMC_ENV_DEV         0
+
+#ifdef CONFIG_POLICY_OTA
+#define CONFIG_SYS_PFE_FW_LENGTH	0x10000
+#define CONFIG_SYS_PFE_FW_ADDR		0xa00000
+#endif
+#endif
 
 /*
  * I2C IO expander
diff --git a/include/configs/ls1046afrwy.h b/include/configs/ls1046afrwy.h
index cff85cb..727572d 100644
--- a/include/configs/ls1046afrwy.h
+++ b/include/configs/ls1046afrwy.h
@@ -177,6 +177,9 @@
 #define CONFIG_ENV_SIZE			0x2000		/* 8KB */
 #define CONFIG_ENV_OFFSET		0x500000	/* 5MB */
 #define CONFIG_ENV_SECT_SIZE		0x40000		/* 256KB */
+#define CONFIG_SYS_FSL_QSPI_BASE	0x40000000
+#define CONFIG_ENV_ADDR			(CONFIG_SYS_FSL_QSPI_BASE + \
+					 CONFIG_ENV_OFFSET)
 #else
 #if defined(CONFIG_SD_BOOT)
 #define CONFIG_SYS_MMC_ENV_DEV		0
-- 
2.7.4

